File: topics/jobs/finom/code/eval-harness.ts
/**
 * Financial Agent Evaluation Harness
 *
 * Demonstrates production-grade evaluation patterns for accounting AI:
 * - Field-level accuracy with severity weighting
 * - Confidence calibration assessment
 * - Per-market breakdown
 * - Regression detection against a baseline
 *
 * This is the evaluation infrastructure an AI team needs before
 * trusting any categorization or booking agent in production.
 *
 * Run with: bun run eval
 */

import { z } from "zod";

// ─── Test Case Schema ────────────────────────────────────────

const TestCase = z.object({
  id: z.string(),
  market: z.enum(["DE", "FR"]),
  description: z.string(),
  severity: z.enum(["critical", "high", "medium", "low"]),
  input: z.object({
    merchant: z.string(),
    amount: z.number(),
    description: z.string(),
    isB2B: z.boolean().default(false),
    counterpartyVatId: z.string().optional(),
  }),
  expected: z.object({
    accountCode: z.string(),
    vatRate: z.number(),
    vatAmount: z.number(),
    netAmount: z.number(),
    mechanism: z.enum(["standard", "reduced", "reverse_charge", "exempt"]),
  }),
});

type TestCase = z.infer<typeof TestCase>;

// ─── Agent Output Schema ─────────────────────────────────────

interface AgentOutput {
  accountCode: string;
  vatRate: number;
  vatAmount: number;
  netAmount: number;
  mechanism: string;
  confidence: number;
}

// ─── Field-Level Comparison ──────────────────────────────────

interface FieldResult {
  field: string;
  expected: string | number;
  actual: string | number;
  correct: boolean;
  delta?: number; // for numeric fields
}

interface CaseResult {
  testCaseId: string;
  market: string;
  severity: string;
  passed: boolean;
  fields: FieldResult[];
  confidence: number;
  severityWeight: number;
  weightedScore: number;
}

function compareFields(expected: TestCase["expected"], actual: AgentOutput): FieldResult[] {
  const results: FieldResult[] = [];

  results.push({
    field: "accountCode",
    expected: expected.accountCode,
    actual: actual.accountCode,
    correct: expected.accountCode === actual.accountCode,
  });

  results.push({
    field: "vatRate",
    expected: expected.vatRate,
    actual: actual.vatRate,
    correct: Math.abs(expected.vatRate - actual.vatRate) < 0.001,
    delta: Math.abs(expected.vatRate - actual.vatRate),
  });

  results.push({
    field: "vatAmount",
    expected: expected.vatAmount,
    actual: actual.vatAmount,
    correct: Math.abs(expected.vatAmount - actual.vatAmount) < 0.01,
    delta: Math.abs(expected.vatAmount - actual.vatAmount),
  });

  results.push({
    field: "netAmount",
    expected: expected.netAmount,
    actual: actual.netAmount,
    correct: Math.abs(expected.netAmount - actual.netAmount) < 0.01,
    delta: Math.abs(expected.netAmount - actual.netAmount),
  });

  results.push({
    field: "mechanism",
    expected: expected.mechanism,
    actual: actual.mechanism,
    correct: expected.mechanism === actual.mechanism,
  });

  return results;
}

// ─── Severity Weighting ──────────────────────────────────────

const SEVERITY_WEIGHTS: Record<string, number> = {
  critical: 4,
  high: 3,
  medium: 2,
  low: 1,
};

// ─── Calibration ─────────────────────────────────────────────

interface CalibrationBin {
  range: [number, number];
  count: number;
  correct: number;
  accuracy: number;
}

function computeCalibration(results: CaseResult[]): {
  bins: CalibrationBin[];
  expectedCalibrationError: number;
} {
  const binEdges = [0, 0.2, 0.4, 0.6, 0.8, 1.0];
  const bins: CalibrationBin[] = [];

  for (let i = 0; i < binEdges.length - 1; i++) {
    const low = binEdges[i];
    const high = binEdges[i + 1];
    const inBin = results.filter(
      (r) => r.confidence >= low && r.confidence < (i === binEdges.length - 2 ? high + 0.01 : high)
    );
    const correct = inBin.filter((r) => r.passed).length;
    bins.push({
      range: [low, high],
      count: inBin.length,
      correct,
      accuracy: inBin.length > 0 ? correct / inBin.length : 0,
    });
  }

  // Expected Calibration Error (ECE)
  const total = results.length;
  const ece = bins.reduce((sum, bin) => {
    if (bin.count === 0) return sum;
    const midpoint = (bin.range[0] + bin.range[1]) / 2;
    return sum + (bin.count / total) * Math.abs(bin.accuracy - midpoint);
  }, 0);

  return { bins, expectedCalibrationError: ece };
}

// ─── Report ──────────────────────────────────────────────────

interface EvalReport {
  totalCases: number;
  overallAccuracy: number;
  severityWeightedAccuracy: number;
  perMarket: Record<string, { accuracy: number; count: number; criticalErrors: number }>;
  perField: Record<string, { accuracy: number; count: number }>;
  calibration: ReturnType<typeof computeCalibration>;
  failures: Array<{ testCaseId: string; failedFields: string[]; severity: string }>;
}

function generateReport(results: CaseResult[]): EvalReport {
  const totalCases = results.length;
  const passed = results.filter((r) => r.passed).length;

  // Severity-weighted accuracy
  const totalWeight = results.reduce((s, r) => s + r.severityWeight, 0);
  const weightedCorrect = results
    .filter((r) => r.passed)
    .reduce((s, r) => s + r.severityWeight, 0);

  // Per-market
  const markets = [...new Set(results.map((r) => r.market))];
  const perMarket: EvalReport["perMarket"] = {};
  for (const m of markets) {
    const inMarket = results.filter((r) => r.market === m);
    const correct = inMarket.filter((r) => r.passed).length;
    const criticalErrors = inMarket.filter(
      (r) => !r.passed && r.severity === "critical"
    ).length;
    perMarket[m] = { accuracy: correct / inMarket.length, count: inMarket.length, criticalErrors };
  }

  // Per-field
  const allFields = results.flatMap((r) => r.fields);
  const fieldNames = [...new Set(allFields.map((f) => f.field))];
  const perField: EvalReport["perField"] = {};
  for (const f of fieldNames) {
    const fieldResults = allFields.filter((fr) => fr.field === f);
    const correct = fieldResults.filter((fr) => fr.correct).length;
    perField[f] = { accuracy: correct / fieldResults.length, count: fieldResults.length };
  }

  // Failures
  const failures = results
    .filter((r) => !r.passed)
    .map((r) => ({
      testCaseId: r.testCaseId,
      failedFields: r.fields.filter((f) => !f.correct).map((f) => f.field),
      severity: r.severity,
    }));

  return {
    totalCases,
    overallAccuracy: passed / totalCases,
    severityWeightedAccuracy: totalWeight > 0 ? weightedCorrect / totalWeight : 0,
    perMarket,
    perField,
    calibration: computeCalibration(results),
    failures,
  };
}

// ─── Simulated Agent Under Test ──────────────────────────────

function simulatedAgent(tc: TestCase): AgentOutput {
  // Simulates an agent that is mostly correct but has realistic failure modes
  const expected = tc.expected;

  // Introduce realistic errors for specific test cases
  if (tc.id === "TC-REVERSE-CHARGE") {
    // Agent misses reverse charge — common real-world failure
    return {
      accountCode: expected.accountCode,
      vatRate: 0.19, // Wrong: applied standard instead of reverse charge
      vatAmount: Math.round((tc.input.amount * 0.19 / 1.19) * 100) / 100,
      netAmount: Math.round((tc.input.amount / 1.19) * 100) / 100,
      mechanism: "standard", // Wrong
      confidence: 0.72,
    };
  }

  if (tc.id === "TC-MIXED-VAT") {
    // Agent gets the category wrong
    return {
      accountCode: "4930", // Wrong code
      vatRate: expected.vatRate,
      vatAmount: expected.vatAmount,
      netAmount: expected.netAmount,
      mechanism: expected.mechanism,
      confidence: 0.55,
    };
  }

  // Default: correct with high confidence
  return {
    ...expected,
    confidence: 0.92 + Math.random() * 0.05,
  };
}

// ─── Test Suite ──────────────────────────────────────────────

const TEST_SUITE: TestCase[] = [
  {
    id: "TC-OFFICE-DE",
    market: "DE",
    description: "Standard office supplies purchase, domestic",
    severity: "low",
    input: {
      merchant: "Büro Discount GmbH",
      amount: 119.0,
      description: "Office supplies",
      isB2B: false,
    },
    expected: {
      accountCode: "4930",
      vatRate: 0.19,
      vatAmount: 19.0,
      netAmount: 100.0,
      mechanism: "standard",
    },
  },
  {
    id: "TC-SOFTWARE-DE",
    market: "DE",
    description: "SaaS subscription, domestic",
    severity: "low",
    input: {
      merchant: "GitHub Inc",
      amount: 47.6,
      description: "GitHub Team monthly",
      isB2B: false,
    },
    expected: {
      accountCode: "4964",
      vatRate: 0.19,
      vatAmount: 7.6,
      netAmount: 40.0,
      mechanism: "standard",
    },
  },
  {
    id: "TC-REVERSE-CHARGE",
    market: "DE",
    description: "Cross-border B2B with VAT ID — reverse charge applies",
    severity: "critical",
    input: {
      merchant: "Dutch Consulting BV",
      amount: 5000.0,
      description: "Consulting services Q1",
      isB2B: true,
      counterpartyVatId: "NL123456789B01",
    },
    expected: {
      accountCode: "4900",
      vatRate: 0,
      vatAmount: 0,
      netAmount: 5000.0,
      mechanism: "reverse_charge",
    },
  },
  {
    id: "TC-TRAVEL-FR",
    market: "FR",
    description: "Train ticket, domestic France",
    severity: "medium",
    input: {
      merchant: "SNCF",
      amount: 120.0,
      description: "Train Paris-Lyon",
      isB2B: false,
    },
    expected: {
      accountCode: "6251",
      vatRate: 0.1,
      vatAmount: 10.91,
      netAmount: 109.09,
      mechanism: "reduced",
    },
  },
  {
    id: "TC-RESTAURANT-DE",
    market: "DE",
    description: "Business dinner, domestic",
    severity: "low",
    input: {
      merchant: "Steakhaus Berlin",
      amount: 178.5,
      description: "Business dinner with client",
      isB2B: false,
    },
    expected: {
      accountCode: "4650",
      vatRate: 0.19,
      vatAmount: 28.5,
      netAmount: 150.0,
      mechanism: "standard",
    },
  },
  {
    id: "TC-MIXED-VAT",
    market: "DE",
    description: "Receipt with standard and reduced VAT items",
    severity: "high",
    input: {
      merchant: "Supermarkt GmbH",
      amount: 53.5,
      description: "Office snacks and supplies",
      isB2B: false,
    },
    expected: {
      accountCode: "4650",
      vatRate: 0.19,
      vatAmount: 8.55,
      netAmount: 44.95,
      mechanism: "standard",
    },
  },
  {
    id: "TC-TELECOM-FR",
    market: "FR",
    description: "Mobile plan, domestic France",
    severity: "low",
    input: {
      merchant: "Orange France",
      amount: 35.99,
      description: "Forfait mobile pro",
      isB2B: false,
    },
    expected: {
      accountCode: "626",
      vatRate: 0.2,
      vatAmount: 6.0,
      netAmount: 29.99,
      mechanism: "standard",
    },
  },
  {
    id: "TC-EXEMPT-DE",
    market: "DE",
    description: "Kleinunternehmer §19 — no VAT charged",
    severity: "critical",
    input: {
      merchant: "Freelancer Schmidt",
      amount: 800.0,
      description: "Design work — Kleinunternehmer",
      isB2B: true,
    },
    expected: {
      accountCode: "4900",
      vatRate: 0,
      vatAmount: 0,
      netAmount: 800.0,
      mechanism: "exempt",
    },
  },
];

// ─── Run Evaluation ──────────────────────────────────────────

function runEvaluation(): EvalReport {
  const results: CaseResult[] = [];

  for (const tc of TEST_SUITE) {
    const output = simulatedAgent(tc);
    const fields = compareFields(tc.expected, output);
    const passed = fields.every((f) => f.correct);
    const severityWeight = SEVERITY_WEIGHTS[tc.severity];

    results.push({
      testCaseId: tc.id,
      market: tc.market,
      severity: tc.severity,
      passed,
      fields,
      confidence: output.confidence,
      severityWeight,
      weightedScore: passed ? severityWeight : 0,
    });
  }

  return generateReport(results);
}

// ─── Main ────────────────────────────────────────────────────

const report = runEvaluation();

console.log("=== Financial Agent Evaluation Report ===\n");
console.log(`Total test cases: ${report.totalCases}`);
console.log(`Overall accuracy: ${(report.overallAccuracy * 100).toFixed(1)}%`);
console.log(`Severity-weighted accuracy: ${(report.severityWeightedAccuracy * 100).toFixed(1)}%`);

console.log("\n--- Per-Market Breakdown ---");
for (const [market, stats] of Object.entries(report.perMarket)) {
  console.log(
    `  ${market}: ${(stats.accuracy * 100).toFixed(1)}% (${stats.count} cases, ${stats.criticalErrors} critical errors)`
  );
}

console.log("\n--- Per-Field Accuracy ---");
for (const [field, stats] of Object.entries(report.perField)) {
  console.log(`  ${field}: ${(stats.accuracy * 100).toFixed(1)}% (${stats.count} comparisons)`);
}

console.log("\n--- Calibration ---");
console.log(`  ECE: ${(report.calibration.expectedCalibrationError * 100).toFixed(2)}%`);
for (const bin of report.calibration.bins) {
  if (bin.count > 0) {
    console.log(
      `  [${bin.range[0].toFixed(1)}-${bin.range[1].toFixed(1)}]: accuracy=${(bin.accuracy * 100).toFixed(0)}%, count=${bin.count}`
    );
  }
}

if (report.failures.length > 0) {
  console.log("\n--- Failures ---");
  for (const f of report.failures) {
    console.log(`  ${f.testCaseId} [${f.severity}]: failed on ${f.failedFields.join(", ")}`);
  }
}

console.log("\n=== Evaluation Complete ===");

File: topics/jobs/finom/code/python-pipeline/eval_harness.py
"""
Financial AI Workflow — Python Evaluation Harness

Demonstrates production-grade evaluation patterns for accounting AI:
- Field-level accuracy with severity weighting
- Confidence calibration (ECE)
- Per-market breakdown
- Regression detection against a stored baseline
- Threshold sweep to find the right auto-book cutoff

This is the eval infrastructure a central AI team owns.
Run it in CI on every model, prompt, or pipeline change.

Run:
    python3 eval_harness.py          # full eval against mock agent
    python3 eval_harness.py --show-details  # include per-case breakdown
"""

import argparse
import math
from dataclasses import dataclass, field
from typing import Literal, Optional

# Import the pipeline under test
from pipeline import (
    Transaction, RoutingStatus,
    categorize, calculate_vat, route,
)


# ─────────────────────────────────────────
# Schema: test case + agent output
# ─────────────────────────────────────────

Severity = Literal["critical", "high", "medium", "low"]
Mechanism = Literal["standard", "reduced", "reverse_charge", "exempt"]

SEVERITY_WEIGHTS: dict[Severity, int] = {
    "critical": 4,
    "high": 3,
    "medium": 2,
    "low": 1,
}


@dataclass
class ExpectedOutput:
    account_code: str
    vat_rate: float
    vat_amount: float
    net_amount: float
    mechanism: Mechanism


@dataclass
class TestCase:
    id: str
    market: str
    description: str
    severity: Severity
    merchant: str
    amount: float
    tx_description: str
    is_b2b: bool
    expected: ExpectedOutput


@dataclass
class FieldResult:
    field: str
    expected: str | float
    actual: str | float
    correct: bool
    delta: Optional[float] = None  # numeric fields only


@dataclass
class CaseResult:
    test_case_id: str
    market: str
    severity: Severity
    passed: bool
    fields: list[FieldResult]
    confidence: float
    severity_weight: int
    weighted_score: float  # severity_weight if passed else 0


# ─────────────────────────────────────────
# Test suite
# ─────────────────────────────────────────

TEST_CASES: list[TestCase] = [
    TestCase(
        id="TC-OFFICE-DE",
        market="DE",
        description="Standard office supply purchase — standard VAT, SKR03",
        severity="medium",
        merchant="Office World",
        amount=119.0,
        tx_description="stationery and printer paper",
        is_b2b=False,
        expected=ExpectedOutput(
            account_code="4930",
            vat_rate=0.19,
            vat_amount=round(119.0 * 0.19 / 1.19, 2),
            net_amount=round(119.0 - 119.0 * 0.19 / 1.19, 2),
            mechanism="standard",
        ),
    ),
    TestCase(
        id="TC-AWS-DE",
        market="DE",
        description="Cloud infrastructure — IT cost account, standard VAT",
        severity="medium",
        merchant="AWS",
        amount=238.0,
        tx_description="EC2 instance monthly charges",
        is_b2b=False,
        expected=ExpectedOutput(
            account_code="4940",
            vat_rate=0.19,
            vat_amount=round(238.0 * 0.19 / 1.19, 2),
            net_amount=round(238.0 - 238.0 * 0.19 / 1.19, 2),
            mechanism="standard",
        ),
    ),
    TestCase(
        id="TC-RESTAURANT-DE",
        market="DE",
        description="Meal expense — reduced VAT applies (7%)",
        severity="high",
        merchant="Lieferando",
        amount=35.70,
        tx_description="team lunch delivery",
        is_b2b=False,
        expected=ExpectedOutput(
            account_code="4650",
            vat_rate=0.07,
            vat_amount=round(35.70 * 0.07 / 1.07, 2),
            net_amount=round(35.70 - 35.70 * 0.07 / 1.07, 2),
            mechanism="reduced",
        ),
    ),
    TestCase(
        id="TC-REVERSE-CHARGE",
        market="DE",
        description="B2B intra-EU service — reverse charge §13b UStG (CRITICAL: wrong here = penalty)",
        severity="critical",
        merchant="Acme Consulting Ltd",
        amount=5000.0,
        tx_description="advisory services",
        is_b2b=True,
        expected=ExpectedOutput(
            account_code="6825",  # professional fees
            vat_rate=0.0,
            vat_amount=0.0,
            net_amount=5000.0,
            mechanism="reverse_charge",
        ),
    ),
    TestCase(
        id="TC-TRAVEL-DE",
        market="DE",
        description="Business travel — Uber ride, reduced VAT",
        severity="medium",
        merchant="Uber",
        amount=45.0,
        tx_description="airport transfer",
        is_b2b=False,
        expected=ExpectedOutput(
            account_code="4670",
            vat_rate=0.07,
            vat_amount=round(45.0 * 0.07 / 1.07, 2),
            net_amount=round(45.0 - 45.0 * 0.07 / 1.07, 2),
            mechanism="reduced",
        ),
    ),
    TestCase(
        id="TC-TRAVEL-FR",
        market="FR",
        description="Business travel in France — PCG code, FR reduced VAT (10%)",
        severity="medium",
        merchant="Uber",
        amount=48.0,
        tx_description="business travel Paris",
        is_b2b=False,
        expected=ExpectedOutput(
            account_code="6251",
            vat_rate=0.10,
            vat_amount=round(48.0 * 0.10 / 1.10, 2),
            net_amount=round(48.0 - 48.0 * 0.10 / 1.10, 2),
            mechanism="reduced",
        ),
    ),
    TestCase(
        id="TC-LUFTHANSA-DE",
        market="DE",
        description="Airline ticket — business travel, reduced VAT",
        severity="medium",
        merchant="Lufthansa",
        amount=320.0,
        tx_description="Frankfurt to Amsterdam flight",
        is_b2b=False,
        expected=ExpectedOutput(
            account_code="4670",
            vat_rate=0.07,
            vat_amount=round(320.0 * 0.07 / 1.07, 2),
            net_amount=round(320.0 - 320.0 * 0.07 / 1.07, 2),
            mechanism="reduced",
        ),
    ),
    TestCase(
        id="TC-LEGAL-DE",
        market="DE",
        description="Legal/notary fees — professional services, standard VAT",
        severity="high",
        merchant="Notary Berlin GbR",
        amount=595.0,
        tx_description="notarial deed",
        is_b2b=False,
        expected=ExpectedOutput(
            account_code="6825",
            vat_rate=0.19,
            vat_amount=round(595.0 * 0.19 / 1.19, 2),
            net_amount=round(595.0 - 595.0 * 0.19 / 1.19, 2),
            mechanism="standard",
        ),
    ),
]


# ─────────────────────────────────────────
# Agent under test: wraps the pipeline
# ─────────────────────────────────────────

@dataclass
class AgentOutput:
    account_code: str
    vat_rate: float
    vat_amount: float
    net_amount: float
    mechanism: str
    confidence: float


def run_agent(tc: TestCase) -> AgentOutput:
    """
    Runs the pipeline under test against a test case.
    In CI, swap this for an API call to the real agent.
    """
    tx = Transaction(
        id=tc.id,
        merchant=tc.merchant,
        amount=tc.amount,
        description=tc.tx_description,
        is_b2b=tc.is_b2b,
        market=tc.market,
    )
    category = categorize(tx)
    vat = calculate_vat(category, tx)
    return AgentOutput(
        account_code=category.account_code,
        vat_rate=vat.rate,
        vat_amount=vat.amount,
        net_amount=vat.net_amount,
        mechanism=vat.mechanism,
        confidence=category.confidence,
    )


# ─────────────────────────────────────────
# Field-level comparison
# ─────────────────────────────────────────

def compare_fields(expected: ExpectedOutput, actual: AgentOutput) -> list[FieldResult]:
    results = []

    results.append(FieldResult(
        field="account_code",
        expected=expected.account_code,
        actual=actual.account_code,
        correct=(expected.account_code == actual.account_code),
    ))
    results.append(FieldResult(
        field="vat_rate",
        expected=expected.vat_rate,
        actual=actual.vat_rate,
        correct=abs(expected.vat_rate - actual.vat_rate) < 0.001,
        delta=abs(expected.vat_rate - actual.vat_rate),
    ))
    results.append(FieldResult(
        field="vat_amount",
        expected=expected.vat_amount,
        actual=actual.vat_amount,
        correct=abs(expected.vat_amount - actual.vat_amount) < 0.01,
        delta=abs(expected.vat_amount - actual.vat_amount),
    ))
    results.append(FieldResult(
        field="net_amount",
        expected=expected.net_amount,
        actual=actual.net_amount,
        correct=abs(expected.net_amount - actual.net_amount) < 0.01,
        delta=abs(expected.net_amount - actual.net_amount),
    ))
    results.append(FieldResult(
        field="mechanism",
        expected=expected.mechanism,
        actual=actual.mechanism,
        correct=(expected.mechanism == actual.mechanism),
    ))
    return results


def evaluate_case(tc: TestCase) -> CaseResult:
    actual = run_agent(tc)
    fields = compare_fields(tc.expected, actual)
    passed = all(f.correct for f in fields)
    weight = SEVERITY_WEIGHTS[tc.severity]
    return CaseResult(
        test_case_id=tc.id,
        market=tc.market,
        severity=tc.severity,
        passed=passed,
        fields=fields,
        confidence=actual.confidence,
        severity_weight=weight,
        weighted_score=weight if passed else 0,
    )


# ─────────────────────────────────────────
# Confidence calibration (ECE)
# ─────────────────────────────────────────

@dataclass
class CalibrationBin:
    low: float
    high: float
    count: int
    correct: int
    accuracy: float


def compute_calibration(results: list[CaseResult]) -> tuple[list[CalibrationBin], float]:
    """
    ECE (Expected Calibration Error): weighted average gap between
    predicted confidence and observed accuracy per bin.
    Target: ECE < 0.05 before widening auto-book thresholds.
    """
    edges = [0.0, 0.2, 0.4, 0.6, 0.8, 1.01]
    bins: list[CalibrationBin] = []

    for i in range(len(edges) - 1):
        low, high = edges[i], edges[i + 1]
        in_bin = [r for r in results if low <= r.confidence < high]
        correct = sum(1 for r in in_bin if r.passed)
        accuracy = correct / len(in_bin) if in_bin else 0.0
        bins.append(CalibrationBin(
            low=low, high=min(high, 1.0),
            count=len(in_bin), correct=correct, accuracy=accuracy,
        ))

    total = len(results)
    ece = sum(
        (b.count / total) * abs(b.accuracy - (b.low + b.high) / 2)
        for b in bins if b.count > 0
    ) if total > 0 else 0.0

    return bins, ece


# ─────────────────────────────────────────
# Regression detection
# ─────────────────────────────────────────

# Simulates a stored baseline from a previous eval run (e.g., CI artifact)
BASELINE: dict[str, str] = {
    "TC-OFFICE-DE": "pass",
    "TC-AWS-DE": "pass",
    "TC-RESTAURANT-DE": "pass",
    "TC-REVERSE-CHARGE": "pass",
    "TC-TRAVEL-DE": "pass",
    "TC-TRAVEL-FR": "pass",
    "TC-LUFTHANSA-DE": "pass",
    "TC-LEGAL-DE": "pass",
}


@dataclass
class Regression:
    test_case_id: str
    severity: Severity
    baseline: str
    current: str


def detect_regressions(results: list[CaseResult]) -> list[Regression]:
    regressions = []
    for r in results:
        baseline = BASELINE.get(r.test_case_id)
        if baseline is None:
            continue
        current = "pass" if r.passed else "fail"
        if baseline == "pass" and current == "fail":
            regressions.append(Regression(
                test_case_id=r.test_case_id,
                severity=r.severity,
                baseline=baseline,
                current=current,
            ))
    return regressions


# ─────────────────────────────────────────
# Threshold sweep
# ─────────────────────────────────────────

@dataclass
class ThresholdAnalysis:
    threshold: float
    auto_book_count: int
    review_count: int
    auto_book_accuracy: float
    critical_errors_in_auto: int


def analyze_thresholds(results: list[CaseResult]) -> list[ThresholdAnalysis]:
    """
    Sweeps auto-book threshold to find the right precision/coverage tradeoff.
    Key invariant: critical errors in auto-book path must be zero.
    """
    analysis = []
    for threshold in [0.70, 0.75, 0.80, 0.85, 0.90, 0.95]:
        auto = [r for r in results if r.confidence >= threshold]
        review = [r for r in results if r.confidence < threshold]
        auto_correct = sum(1 for r in auto if r.passed)
        critical_in_auto = sum(
            1 for r in auto if not r.passed and r.severity == "critical"
        )
        analysis.append(ThresholdAnalysis(
            threshold=threshold,
            auto_book_count=len(auto),
            review_count=len(review),
            auto_book_accuracy=auto_correct / len(auto) if auto else 0.0,
            critical_errors_in_auto=critical_in_auto,
        ))
    return analysis


# ─────────────────────────────────────────
# Report
# ─────────────────────────────────────────

def print_report(
    results: list[CaseResult],
    show_details: bool = False,
) -> None:
    total = len(results)
    passed = sum(1 for r in results if r.passed)
    total_weight = sum(r.severity_weight for r in results)
    weighted_correct = sum(r.weighted_score for r in results)

    # Per-market
    markets = sorted({r.market for r in results})
    per_market = {}
    for m in markets:
        in_m = [r for r in results if r.market == m]
        per_market[m] = {
            "count": len(in_m),
            "accuracy": sum(1 for r in in_m if r.passed) / len(in_m),
            "critical_errors": sum(1 for r in in_m if not r.passed and r.severity == "critical"),
        }

    # Per-field
    all_fields = [f for r in results for f in r.fields]
    field_names = list(dict.fromkeys(f.field for f in all_fields))
    per_field = {}
    for fn in field_names:
        subset = [f for f in all_fields if f.field == fn]
        per_field[fn] = sum(1 for f in subset if f.correct) / len(subset)

    bins, ece = compute_calibration(results)
    regressions = detect_regressions(results)
    threshold_analysis = analyze_thresholds(results)
    failures = [r for r in results if not r.passed]

    # ── Print ──
    print("\n══════════════════════════════════════════════════════════")
    print("  Financial AI Workflow — Evaluation Report")
    print("══════════════════════════════════════════════════════════\n")

    print(f"  Total cases:              {total}")
    print(f"  Overall accuracy:         {passed}/{total}  ({passed/total:.0%})")
    swa = weighted_correct / total_weight if total_weight else 0
    print(f"  Severity-weighted acc:    {swa:.2%}  ({'✓ PASS' if swa >= 0.90 else '✗ FAIL'})")
    print(f"  ECE (calibration):        {ece:.4f}  ({'✓ < 0.05' if ece < 0.05 else '✗ > 0.05 — threshold unreliable'})\n")

    print("  ── Per-market ─────────────────────────────────────────")
    for m, stats in per_market.items():
        critical_tag = f"  ← {stats['critical_errors']} CRITICAL ERROR(s)" if stats["critical_errors"] else ""
        print(f"  {m}:  {stats['accuracy']:.0%} accuracy ({stats['count']} cases){critical_tag}")

    print("\n  ── Per-field accuracy ─────────────────────────────────")
    for fn, acc in per_field.items():
        marker = "✓" if acc >= 0.90 else "✗"
        print(f"  {marker} {fn:<15} {acc:.0%}")

    print("\n  ── Confidence calibration (ECE) ───────────────────────")
    for b in bins:
        if b.count == 0:
            continue
        midpoint = (b.low + b.high) / 2
        gap = abs(b.accuracy - midpoint)
        bar = "█" * round(b.accuracy * 20)
        print(f"  [{b.low:.1f}–{b.high:.1f}] n={b.count:<3} acc={b.accuracy:.0%} conf≈{midpoint:.0%}  Δ={gap:.2f}  {bar}")

    print("\n  ── Threshold sweep (auto-book tradeoff) ───────────────")
    print(f"  {'Threshold':<12} {'Auto-book':<12} {'Review':<10} {'Auto-acc':<12} {'Crit-errors'}")
    for t in threshold_analysis:
        crit_flag = "  ← BLOCK" if t.critical_errors_in_auto > 0 else ""
        recommended = "  ← recommended" if t.threshold == 0.85 and t.critical_errors_in_auto == 0 else ""
        print(f"  {t.threshold:<12.2f} {t.auto_book_count:<12} {t.review_count:<10} {t.auto_book_accuracy:<12.0%} {t.critical_errors_in_auto}{crit_flag}{recommended}")

    if regressions:
        print("\n  ── ✗ REGRESSIONS DETECTED ─────────────────────────────")
        for reg in regressions:
            print(f"  [{reg.severity.upper()}] {reg.test_case_id}: was {reg.baseline}, now {reg.current}")
    else:
        print("\n  ── ✓ No regressions vs baseline ───────────────────────")

    if failures:
        print("\n  ── Failures ───────────────────────────────────────────")
        for r in failures:
            bad_fields = [f.field for f in r.fields if not f.correct]
            print(f"  [{r.severity.upper()}] {r.test_case_id}: failed fields → {', '.join(bad_fields)}")

    if show_details:
        print("\n  ── Per-case detail ────────────────────────────────────")
        for r in results:
            status = "✓" if r.passed else "✗"
            print(f"  {status} [{r.severity.upper():<8}] {r.test_case_id} (conf={r.confidence:.0%})")
            for f in r.fields:
                field_ok = "  " if f.correct else "✗ "
                delta_str = f"  Δ={f.delta:.4f}" if f.delta and not f.correct else ""
                if not f.correct:
                    print(f"       {field_ok}{f.field}: expected={f.expected}  got={f.actual}{delta_str}")

    print("\n══════════════════════════════════════════════════════════\n")

    # CI exit code: fail if regressions or critical errors
    has_critical = any(
        not r.passed and r.severity == "critical" for r in results
    )
    if regressions or has_critical:
        print("  ✗ EVAL FAILED — regression or critical error detected")
        print("  This run would block a CI pipeline.\n")
    else:
        print("  ✓ EVAL PASSED — no regressions, no critical errors\n")


# ─────────────────────────────────────────
# Entry point
# ─────────────────────────────────────────

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--show-details", action="store_true", help="Show per-case field breakdown")
    args = parser.parse_args()

    results = [evaluate_case(tc) for tc in TEST_CASES]
    print_report(results, show_details=args.show_details)
